svm/src/transaction_processor.rs

```diff
@@ -309,62 +298,109 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
+                &mut account_loader,
+                tx,
+                validate_result,
+                &mut error_metrics,
+                &environment.feature_set,
+                environment
+                    .rent_collector
+                    .unwrap_or(&RentCollector::default()),
+            ));
+            load_us = load_us.saturating_add(single_load_us);
+
+            let (processing_result, single_execution_us) = measure_us!(match load_result {
+                TransactionLoadResult::NotLoaded(err) => Err(err),
+                TransactionLoadResult::FeesOnly(fees_only_tx) => {
+                    if enable_transaction_loading_failure_fees {
+                        // Update loaded accounts cache with nonce and fee-payer
+                        account_loader
+                            .update_accounts_for_failed_tx(tx, &fees_only_tx.rollback_accounts);
+
+                        Ok(ProcessedTransaction::FeesOnly(Box::new(fees_only_tx)))
+                    } else {
+                        Err(fees_only_tx.load_error)
                     }
-                    TransactionLoadResult::Loaded(loaded_transaction) => {
-                        let executed_tx = self.execute_loaded_transaction(
-                            tx,
-                            loaded_transaction,
-                            &mut execute_timings,
-                            &mut error_metrics,
-                            &mut program_cache_for_tx_batch,
-                            environment,
-                            config,
-                        );
-
-                        // Update batch specific cache of the loaded programs with the modifications
-                        // made by the transaction, if it executed successfully.
-                        if executed_tx.was_successful() {
-                            program_cache_for_tx_batch.merge(&executed_tx.programs_modified_by_tx);
-                        }
+                }
+                TransactionLoadResult::Loaded(loaded_transaction) => {
+                    let executed_tx = self.execute_loaded_transaction(
+                        tx,
+                        loaded_transaction,
+                        &mut execute_timings,
+                        &mut error_metrics,
+                        &mut account_loader.program_cache,
+                        environment,
+                        config,
+                    );
 
-                        Ok(ProcessedTransaction::Executed(Box::new(executed_tx)))
+                    // Update batch specific cache of the loaded programs with the modifications
+                    // made by the transaction, if it executed successfully.
+                    if executed_tx.was_successful() {
+                        account_loader
+                            .program_cache
+                            .merge(&executed_tx.programs_modified_by_tx);
```

Comment:
I think we also need to remove accounts from the program cache here when they no longer meet the criteria for being added to the program cache. Otherwise you could have a case where different batchings of a list of transactions could be processed differently. Imagine you have a list of transactions where tx `A` closes a buffer account `X` and reallocs the account with a non-loader owner and tx `B` that loads account `X` with an invisible read. If txs `A` and `B` are in the same batch, `B` would use the program cache to get the stale size of account `X` because it was never removed from the program cache. If tx `A` is first processed in one batch and then tx `B` is processed in a subsequent batch, account `X` wouldn't be included in the program cache when processing the batch of txs which includes tx `B` and so it would use the correct size.

Reply:
Rather than trying to prune the program cache here, I think it's probably best to first fix the issues with how the accounts loader uses the program cache in a separate change.. womp womp

Reply:
ill add a test to cover it, but this works correctly. this block comes after execution in the transaction processing loop:
```rust
if executed_tx.was_successful() {
    account_loader
        .program_cache
        .merge(&executed_tx.programs_modified_by_tx);
}
```

which ensures the batch-local program cache is updated after after transaction

Reply:
Can you explain how it works correctly? The loaded account size will be different depending on how transactions are batched up in my example above, won't they?

Reply:
(discussed in slack and i misunderstood the specific issue because it depends on a very subtle kind of usage, should be an easy fix)

Reply:
context for andrew: i thought what starry meant is the case where you close an upgradeable bpf loader buffer and then reopen it under a new owner, do subsequent transactions see the old buffer account? but they do in fact (correctly) see the new account

the actual issue was that if you close the buffer, reopen under a new owner with new size, and then append it to the account keys without using it in an instruction, we go to the program cache and create a dummy `AccountSharedData` from the stale program cache entry, because closed accounts are marked with a tombstone, rather than evicted. so this is another edge case that results from the account loader not checking tombstones (which it cannot do until fee-only transactions are enabled)

also i discovered while working on this that tombstoned accounts always report a size of 0, even if they are existing bpf loader buffers, so transaction size differs here too depending on whether you pass it as an instruction account

thankfully the close/reopen case is easy to handle. since we necessarily have an account cache entry (closing the buffer account cannot be done without passing it to the instruction), we check the owner and only use the cache if owned by a loader:
* if the account was reopened with a non-loader owner, we use the state from account cache normally
* if the account was closed, we use the program cache, and it doesnt matter since both sizes are 0
* if the account was reopened and assigned to a loader then size depends on whether it is an instruction account:
  * an instruction account uses the state from account cache normally
  * a non-instruction account incorrectly reports the size as 0, but this is identical to the current behavior in master, so no changes are required

ive added various tests for close/reopen case and for the simple size difference case

all this is yet another side effect of consensus-affecting program cache misuse, so like the other edge cases, we can fix all this and get rid of all the special handling once fee-only transactions are activated and [SIMD186](https://github.com/solana-foundation/solana-improvement-documents/pull/186) is approved and implemented

(also @jstarry if you would like to give extra eyes on that simd i promise it is much shorter than this pr :smile:)